From 8055553846ea7da254c2d99d4aab07c0de183633 Mon Sep 17 00:00:00 2001
From: Quynh Nguyen <quynhnguyen.xb@renesas.com>
Date: Wed, 7 Oct 2020 13:18:38 +0700
Subject: [PATCH 058/433] ARM: shmobile: Add CCI configuration for r8a7742

Refer work from https://patchwork.kernel.org/patch/3491641/
Add the settings of CCI-400 for enabling ca15/ca7 snoop

Signed-off-by: Gaku Inami <gaku.inami.xw@bp.renesas.com>
Signed-off-by: Magnus Damm <damm@opensource.se>
Signed-off-by: Binh Nguyen <binh.nguyen.uw@rvc.renesas.com>
Signed-off-by: Hao Bui <hao.bui.yg@renesas.com>
Signed-off-by: Quynh Nguyen <quynhnguyen.xb@renesas.com>
---
 arch/arm/mach-shmobile/smp-r8a7790.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/arch/arm/mach-shmobile/smp-r8a7790.c b/arch/arm/mach-shmobile/smp-r8a7790.c
index 4b33d43..7ca4bbb 100644
--- a/arch/arm/mach-shmobile/smp-r8a7790.c
+++ b/arch/arm/mach-shmobile/smp-r8a7790.c
@@ -26,6 +26,12 @@
 #include "rcar-gen2.h"
 #include "r8a7790.h"
 
+#define CCI_BASE	0xf0090000
+#define CCI_SLAVE3	0x4000
+#define CCI_SLAVE4	0x5000
+#define CCI_SNOOP	0x0000
+#define CCI_STATUS	0x000c
+
 static const struct rcar_sysc_ch r8a7790_ca15_scu = {
 	.chan_offs = 0x180, /* PWRSR5 .. PWRER5 */
 	.isr_bit = 12, /* CA15-SCU */
@@ -49,6 +55,8 @@ static struct rcar_apmu_config r8a7790_apmu_config[] = {
 
 static void __init r8a7790_smp_prepare_cpus(unsigned int max_cpus)
 {
+	void __iomem *p;
+
 	/* let APMU code install data related to shmobile_boot_vector */
 	shmobile_smp_apmu_prepare_cpus(max_cpus,
 				       r8a7790_apmu_config,
@@ -58,6 +66,17 @@ static void __init r8a7790_smp_prepare_cpus(unsigned int max_cpus)
 	rcar_gen2_pm_init();
 	rcar_sysc_power_up(&r8a7790_ca15_scu);
 	rcar_sysc_power_up(&r8a7790_ca7_scu);
+
+	/* enable snoop and DVM */
+	p = ioremap_nocache(CCI_BASE, 0x8000);
+	writel_relaxed(readl_relaxed(p + CCI_SLAVE3 + CCI_SNOOP) | 0x3,
+		p + CCI_SLAVE3 + CCI_SNOOP);    /* ca15 */
+	writel_relaxed(readl_relaxed(p + CCI_SLAVE4 + CCI_SNOOP) | 0x3,
+		p + CCI_SLAVE4 + CCI_SNOOP);    /* ca7 */
+	while (__raw_readl(p + CCI_STATUS))
+		;
+	/* wait for pending bit low */
+	iounmap(p);
 }
 
 struct smp_operations r8a7790_smp_ops __initdata = {
-- 
2.7.4

